src/analysis_and_optimization/Memory_patterns.ml

```diff
@@ -195,111 +195,60 @@ and query_initial_demotable_funs (in_loop : bool) (acc : string Set.Poly.t)
   | UserDefined ((_ : string), (_ : bool Fun_kind.suffix)) ->
       Set.Poly.union acc demoted_and_top_level_names
 
-(**
-  Check whether any functions in the right hand side expression of an assignment
-  support SoA. If so then return true, otherwise return false.
- *)
-let rec is_any_soa_supported_expr
-    Expr.Fixed.{pattern; meta= Expr.Typed.Meta.{adlevel; type_; _}} : bool =
-  if
-    UnsizedType.is_dataonlytype adlevel
-    || not (UnsizedType.contains_eigen_type type_)
-  then true
-  else
-    match pattern with
-    | FunApp (kind, (exprs : Expr.Typed.t list)) ->
-        is_any_soa_supported_fun_expr kind exprs
-    | Indexed (expr, (_ : Expr.Typed.t Index.t list))
-     |Promotion (expr, _, _)
-     |TupleProjection (expr, _) ->
-        is_any_soa_supported_expr expr
-    | Var (_ : string) | Lit ((_ : Expr.Fixed.Pattern.litType), (_ : string)) ->
-        true
-    | TernaryIf (_, texpr, fexpr) ->
-        is_any_soa_supported_expr texpr && is_any_soa_supported_expr fexpr
-    | EAnd (lhs, rhs) | EOr (lhs, rhs) ->
-        is_any_soa_supported_expr lhs && is_any_soa_supported_expr rhs
+let print_string s = print_endline s
+let print_set s = Set.Poly.iter ~f:print_string s
 
-(**
-  Return false if the [Fun_kind.t] does not support [SoA]
- *)
-and is_any_soa_supported_fun_expr (kind : 'a Fun_kind.t)
-    (exprs : Expr.Typed.t list) : bool =
-  match kind with
-  | CompilerInternal (Internal_fun.FnMakeArray | FnMakeRowVec) -> false
-  | UserDefined ((_ : string), (_ : bool Fun_kind.suffix)) -> false
-  | CompilerInternal (_ : 'a Internal_fun.t) -> true
-  | Fun_kind.StanLib (name, (_ : bool Fun_kind.suffix), _) -> (
-    match name with
-    | "check_matching_dims" -> true
-    | _ ->
-        is_fun_soa_supported name exprs
-        && List.exists ~f:is_any_soa_supported_expr exprs )
+let print_bool nn bb =
+  Printf.printf "%s: %s \n" nn (if bb then "true" else "false")
 
-(**
-  Return true if the rhs expression of an assignment contains only
-   combinations of AutoDiffable Reals and Data Matrices
- *)
-let rec is_any_ad_real_data_matrix_expr
-    Expr.Fixed.{pattern; meta= Expr.Typed.Meta.{adlevel; _}} : bool =
-  if UnsizedType.is_dataonlytype adlevel then false
-  else
+let rec get_lowest_level
```

Comment:
I think these functions deserve a docstring (and possibly a better name?)

Reply:
Just added doc to explain what the recursion is looking for. And gave these better names